Add in out of core sort

revans2 opened the pull request:
This depends on https://github.com/rapidsai/cudf/pull/7381

The code is functionally complete and does a lot of things.

It improves the memory utilization of non-out of core sorting by not gathering columns that will not be output.
It adds in support for spill metrics so sort can report how many bytes were spilled
It fixes a bug in hash aggregate that made first and last not output the correct answer if multiple batches came in.
and it adds in support for out of core sort.

I still need to do some documentation work for sort.
I have been running some benchmarks and out of core is noticeably slower than the other version, but it can sort massive amounts of data. (I was able to sort 100GB in a single task on a 16GB V100), The performance was also fairly awful, but it was spilling.

There are a few more optimizations I want to try and play around with, but I was also hoping for some feedback/ideas on how to improve the performance.

If the code is hard to follow please let me know so I can add in more comments.

I am really interested in ideas for how to split up the batches to combine later on, because this is key to the algorithms performance. Right now I split incoming batches up into ones that are about `target size/8`.  This means two things. First, if you want to sort data that is smaller than `target size/8` or with default configs that is `2GB/8` = `256MB` then no merge sort will be done. The overhead is in putting the data into the spill buffer and pulling it back out. I need to do some profiling on this and clean up but ideally it should match the original time. The second thing this means is that if we get less than 8 times the target size (16GB) of input data then we can likely sort all of the data in just 2 passes. One pass to sort it and store it, and the second pass to merge sort and output results.

If we could be more dynamic in this we could probably get better numbers in both cases, but I am not sure of a good algorithm to do that.

This addresses part of #1642 it should fix it until we start to support rank, but then we need to look at out of core window functions.

This fixes #19

revans2 commented:
This should now be ready for review.  The dependency was merged into cudf 0.19 this morning, but will probably not be all the way through CI until tomorrow.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala:
```diff
@@ -46,4 +56,211 @@ object SortUtils {
+
+  lazy val computationSchema: Seq[Attribute] =
+    needsComputation.map { so =>
+      AttributeReference(s"SORT_TMP", so.dataType, so.nullable)()
```
I'm not following the use of `SORT_TMP`.  A block comment at the top of the class describing at a high level what the algorithm is and how extra columns are used would help.

jlowe replied:
Would still like to see a block comment on the class describing the high-level algorithm being performed (i.e.: mentioning the fact that there are extra columns, how they're computed and used, maybe walk through an example, etc.)

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
-      childPlans(0).convertIfNeeded())
-
-  override def tagPlanForGpu(): Unit = {
-    if (GpuOverrides.isAnyStringLit(sort.sortOrder)) {
-      willNotWorkOnGpu("string literal values are not supported in a sort")
-    }
```
Does this limitation no longer apply?  I didn't see where this check moved if it does.

revans2 replied:
Yup it didn't apply a while ago, but we never got around to removing it.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
-    Some(exec.gpuLongMetric(metricName))
-  } else {
-    None
+case class OutOfCoreBatch(buffer: SpillableColumnarBatch, row: UnsafeRow) extends AutoCloseable {
```
A one-liner comment explaining what this is for would help.  I get the `buffer` part, `row` is a little less clear when browsing the code.

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+}
 
-      private def closeBatch(): Unit = resultBatch.foreach(_.close())
+case class GpuOutOfCoreSortIterator(
```
A block comment here explaining the high-level algorithm would help

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+      withResource(sortedTbl.contiguousSplit(splitIndexes: _*)) { split =>
+        memUsed += split.map(_.getBuffer.getLength).sum
+        val stillPending = if (sortedOffset >= 0) {
+          withResource(new NvtxRange("Store Sorted Table", NvtxColor.DARK_GREEN)) { _ =>
```
Is this block expected to take a significant amount of time?  I'm surprised to see an NVTX range here, as I'm seeing only some object creation and adding to a linked list.

abellina replied:
I think this range is mainly for visibility.

revans2 replied:
I added it when debugging what was taking a lot of time, and never removed it.  It turns out that the time was being taken by the next range where we add things to pending, because pending will do a lazy compile of the comparator which I was trying to track down. I am happy to remove this one if we want to.

jlowe replied:
Unless there's a compelling reason for them to remain, I'd like to see the extremely small NVTX ranges like this one removed since they can distort profiles and probably aren't that useful in practice.  If these tiny ranges serve a useful purpose, may be worth commenting so someone doesn't come along later and remove them.

abellina replied:
Like with metrics, and log lines, should we have levels for nvtx ranges? This is a separate discussion but since we are talking about it, I can spawn off an issue if that's interesting.

revans2 replied:
If I am looking at NVTX everything is already debug level. The main reason to remove them is because they are so small all the time that showing them has no useful value. I just didn't remove them earlier when I was debugging, and Jason did the right thing to ask "why is this section of code taking long enough that we need to pay attention to it?" Because if it did take that long then we have some optimizing to do.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
 
+sealed trait SortExecType extends Serializable
+
+object OutOfCoreSort extends SortExecType
```
A comment on each of these types would be good. I.e. I am not 100% sure why we wouldn't always want `OutOfCoreSort` though I believe that's to do with unstable sort.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+        val cpuOrd = new LazilyGeneratedOrdering(sorter.cpuOrdering)
+        val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
+        val iter = GpuOutOfCoreSortIterator(cbIter, sorter, cpuOrd,
+          // bad things can happen if the target size is really small
```
are the bad things something documented later? :)

abellina reviewed (changes requested):
I am mostly checking with `GpuSortExec` now.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
-    if (global) OrderedDistribution(sparkSortOrder) :: Nil else UnspecifiedDistribution :: Nil
+    if (global) OrderedDistribution(sortOrder) :: Nil else UnspecifiedDistribution :: Nil
 
   // Eventually this might change, but for now we will produce a single batch, which is the same
```
comment needs updating after the change.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+    // something really complicated and figure out how much data get per batch, but in practice
+    // we really only expect to see one or two batches worth of data come in, so lets optimize
+    // for that case and set the targetBatchSize to always be 1/8th the targetSize.
+    val targetBatchSize = targetSize / 8
```
In the poorly configured case, `targetSize` could be 16KB. So this means we'd do 2KB batches here?

revans2 replied:
Yes, but the only time that would really happen is unit tests.  We eventually might want to prevent people from shooting themselves in the foot, but for some reason I am feeling optimistic about human nature today, and I don't want to have to figure out how to make the tests that set it to 0 still work if we disallow setting it to anything smaller than say a few hundred MB.

abellina replied:
> for some reason I am feeling optimistic about human nature today

:)

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
-      override def hasNext: Boolean = {
-        if (resultBatch.isDefined) {
-          true
+      withResource(sortedTbl.contiguousSplit(splitIndexes: _*)) { split =>
```
```suggestion
      withResource(sortedTbl.contiguousSplit(splitIndexes: _*)) { splits =>
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+          mergedBatch.numRows() - 1
+        } else {
+          // The data is only fully sorted if there is nothing pending that is smaller than it
+          // so get the nest smallest row that is pending
```
```suggestion
          // so get the next smallest row that is pending
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala:
```diff
@@ -16,254 +16,459 @@
+class Pending(cpuOrd: LazilyGeneratedOrdering) extends AutoCloseable {
+  private val pending = new PriorityQueue[OutOfCoreBatch](new Comparator[OutOfCoreBatch]() {
+    override def compare(a: OutOfCoreBatch, b: OutOfCoreBatch): Int =
+      cpuOrd.compare(a.row, b.row)
```
Could `row` be renamed to something like `cutOff` or `lower`, if I understand this correctly?

revans2 commented on docs/configs.md:
```diff
@@ -78,6 +78,7 @@ Name | Description | Default Value
 <a name="sql.incompatibleDateFormats.enabled"></a>spark.rapids.sql.incompatibleDateFormats.enabled|When parsing strings as dates and timestamps in functions like unix_timestamp, setting this to true will force all parsing onto GPU even for formats that can result in incorrect results when parsing invalid inputs.|false
 <a name="sql.incompatibleOps.enabled"></a>spark.rapids.sql.incompatibleOps.enabled|For operations that work, but are not 100% compatible with the Spark equivalent set if they should be enabled by default or disabled by default.|false
 <a name="sql.metrics.level"></a>spark.rapids.sql.metrics.level|GPU plans can produce a lot more metrics than CPU plans do. In very large queries this can sometimes result in going over the max result size limit for the driver. Supported values include DEBUG which will enable all metrics supported and typically only needs to be enabled when debugging the plugin. MODERATE which should output enough ...
+<a name="sql.outOfCoreSort.enabled"></a>spark.rapids.sql.outOfCoreSort.enabled|enable or disable out of core sorting. Out of core sorting allows for sorting large skewed data sets, but is not a stable sort. Apache Spark does not guarantee a stable sort in most cases, except when coalescing all data to a single task and sorting there. In that case we will produce different results.|true
```
I personally think that "out of core" or "external memory" would be the correct term to use. Spillable is very Spark specific, but that is the audience we are going for so if you insist I will change it. The only reason this is configurable is because of that one small use case where the stable sort matters. I have not seen enough of a performance difference to justify turning it off for performance. 

Perhaps we should just call it stable vs unstable sort, and in the future we can add in an out of core stable sort implementation.

jlowe replied:
> if you insist I will change it.

I'm not insisting, just pointing out that the term will not be familiar to many Spark users reading the documentation and it's not really defined anywhere.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExec.scala:
```diff
@@ -57,6 +57,9 @@ object GpuMetric {
   val BUILD_DATA_SIZE = "buildDataSize"
   val BUILD_TIME = "buildTime"
   val STREAM_TIME = "streamTime"
+  val SPILL_AMOUNT = "spillData"
```
Which do you want it to be?  Right now it is all data spilled to host memory or disk.

jlowe replied:
> it is all data spilled to host memory or disk.

That's fine, the description should be updated to say this or something similar so the user knows what is being measured.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuExec.scala:
```diff
@@ -199,6 +223,12 @@ trait GpuExec extends SparkPlan with Arm {
   lazy val additionalMetrics: Map[String, GpuMetric] = Map.empty
 
+  protected def spillMetrics: Map[String, GpuMetric] = Map(
+    SPILL_AMOUNT -> createSizeMetric(ESSENTIAL_LEVEL, DESCRIPTION_SPILL_AMOUNT),
+    SPILL_AMOUNT_DISK -> createSizeMetric(MODERATE_LEVEL, DESCRIPTION_SPILL_AMOUNT_DISK),
+    SPILL_AMOUNT_HOST -> createSizeMetric(MODERATE_LEVEL, DESCRIPTION_SPILL_AMOUNT_HOST)
```
If the GPU OOM spill code already uses them, then wouldn't we be double counting because the the callback also happens because of a GPU OOM.

jlowe replied:
Yeah, I was originally thinking we should just let the OOM spill code handle this, but that doesn't tie it to a specific exec node which we want here.

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBuffer.scala:
```diff
@@ -72,6 +72,11 @@ trait RapidsBuffer extends AutoCloseable {
+  /**
+   * A callback function that can be used for metrics when a buffer is spilled
+   */
+  val spillCallback: (String, Long) => Unit
```
I started out with not putting defaults everywhere because I wanted to be sure that I had plumbed it through properly in all places with compile errors.  If you want me to move it to RapidsBufferId I can do it.

revans2 commented:
build

revans2 commented:
@jlowe and @abellina I think I have addressed most of your review comments. There are still a few things left that I asked questions on. I think I will switch the configs over to talk about stable vs non-stable sorts instead of out of core. But I want to hold off to be sure that there are not other things found/decided on.

revans2 commented:
build

revans2 commented:
@jlowe @abellina I think I have addressed all of your review comments. Please take another look

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBuffer.scala:
```diff
@@ -72,6 +79,19 @@ trait RapidsBuffer extends AutoCloseable {
+   * <li><code>amount</code> the amount of data in bytes that is spilled.</li>
+   * </ul>
+   */
+  val spillCallback: (StorageTier, StorageTier, Long) => Unit
```
I don't default it here because the only time it needs to be defaulted is in the degenerate case and tests. I would rather be cautious and force someone adding a new buffer to decide if they need it or not, but with RapidsBufferBase it may be a moot point.

revans2 commented:
build

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala:
```diff
@@ -46,4 +59,265 @@ object SortUtils {
+ * include sorting, merge sorting, and finding bounds. These can be combined in various ways to
+ * do different algorithms. When you are done with these different operations you can drop the
+ * temporary columns that were added, just for computation, using `removeProjectedColumns`.
+ * Some times you may want to pull data back to the CPU and sort rows there too. we provide
```
```suggestion
 * Some times you may want to pull data back to the CPU and sort rows there too. We provide
```

jlowe reviewed (approved):
One small typo that doesn't really matter.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/SortUtils.scala:
```diff
@@ -46,4 +56,211 @@ object SortUtils {
+
+  lazy val computationSchema: Seq[Attribute] =
+    needsComputation.map { so =>
+      AttributeReference(s"SORT_TMP", so.dataType, so.nullable)()
```
@revans2 this one is a small nit, but I believe it was missed. Not a blocker.

revans2 commented:
I think I got the final nits from everyone.  Thanks for the reviews

revans2 commented:
build
